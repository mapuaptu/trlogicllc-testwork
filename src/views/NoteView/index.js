import Button from '@/components/Button/index.vue';
import Icon from '@/components/Icon/index.vue';
import Todo from '@/components/Todo/index.vue';
import deleteNote from '@/mixins/deleteNote';
import ModalCancelEdit from '@/views/Modals/CancelEdit/index.vue';
import ModalCreateTodo from '@/views/Modals/CreateTodo/index.vue';
import ClickOutside from 'vue-click-outside';

export default {
  name: 'NoteView',
  directives: {
    ClickOutside,
  },
  mixins: [deleteNote],
  components: {
    Button,
    Icon,
    Todo,
  },
  data() {
    return {
      id: '',
      name: '',
      todos: [],
      editMode: false,
    };
  },
  methods: {
    editName() {
      this.editMode = !this.editMode;

      return this.$nextTick(() => {
        this.editMode && this.$refs.input.focus();
      });
    },
    hideEditMode() {
      return (this.editMode = false);
    },
    onNameInput(event) {
      // If name of note is empty - remain first symbol

      if (!event.target.value) {
        return;
      }

      return (this.name = event.target.value);
    },
    onCancel() {
      return this.$modal.show(ModalCancelEdit, {}, { height: 'auto' });
    },
    onSave() {
      this.$store.commit('EDIT_NOTE', {
        name: this.name,
        todos: this.todos,
        id: this.id,
      });

      return this.$router.push({ name: 'home' });
    },
    createTodo() {
      return this.$modal.show(
        ModalCreateTodo,
        {
          // react style becase such modal component
          // issue on github - https://github.com/euvl/vue-js-modal/issues/192#issuecomment-372990690
          onTodoSave: this.onTodoSave,
        },
        { height: 'auto' },
      );
    },
    onTodoSave(name) {
      if (this.todos.find((item) => item.name === name)) {
        return {
          error: { message: 'Todo with same name already exist' },
        };
      }

      // may be use uuid is better

      this.todos.push({ name, complete: false, id: Date.now() });

      return { error: null };
    },
  },
  created() {
    const note = this.$store.state.notes.find(
      (item) => item.id === Number(this.$route.params.id),
    );

    // create new instance of todos object for immutability of original todos

    if (note) {
      this.name = note.name;
      this.id = note.id;
      this.todos = note.todos.map((item) => ({ ...item }));
    }
  },
};
