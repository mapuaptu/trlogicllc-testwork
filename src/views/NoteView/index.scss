@import '@/styles/_variables.scss';

.note {
  position: relative;

  .title {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 40px;
    border-bottom: 1px solid $color-java;
    padding: 10px 20px;

    .controls {
      display: flex;
      align-items: center;

      > div:not(:last-child) {
        margin-right: 15px;
      }
    }

    .text {
      max-width: 40%;
      font-size: 25px;
      line-height: 25px;
      word-break: break-all;
      color: $color-java;
    }

    .input {
      border: none;
      max-width: 40%;
      font-size: 25px;
      line-height: 25px;
      color: $color-java;
      background-color: transparent;
    }
  }

  .content {
    display: flex;
    flex-flow: column;
    padding: 10px 20px;
    min-height: 100px;

    .button {
      align-self: flex-end;
    }
  }

  .footer {
    display: flex;
    justify-content: flex-end;
    border-top: 1px solid $color-java;
    padding: 10px 20px;

    > div:not(:last-child) {
      margin-right: 15px;
    }
  }
}
