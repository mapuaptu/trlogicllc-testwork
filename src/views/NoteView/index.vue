<template>
  <div :class="$style.note">
    <div
      v-click-outside="hideEditMode"
      :class="$style.title"
    >
      <div
        v-if="!editMode"
        :class="$style.text"
      >
        {{ name }}
      </div>

      <input
        v-else
        ref="input"
        :value="name"
        type="text"
        :class="$style.input"
        @input="onNameInput"
      >

      <div :class="$style.controls">
        <Button
          icon="pencil-outline"
          @click.native="editName"
        >
          {{ editMode ? 'Save name': 'Edit name' }}
        </Button>

        <Button
          icon="delete-outline"
          @click.native="deleteNote({id, name})"
        >
          Delete note
        </Button>
      </div>
    </div>

    <div
      v-if="message"
      :class="$style.message"
    >
      {{ message }}
    </div>

    <div :class="$style.content">
      <Button
        :fill="true"
        icon="plus-outline"
        :class="$style.button"
        @click.native="createTodo"
      >
        Add todo
      </Button>

      <div :class="$style.todos">
        <Todo
          v-for="item in todos"
          :key="item.id"
          v-model="item.complete"
          @edit="editTodo(item)"
          @delete="deleteTodo(item)"
        >
          {{ item.name }}
        </Todo>
      </div>
    </div>

    <div :class="$style.footer">
      <Button @click.native="onCancel">
        Cancel
      </Button>

      <Button
        :fill="true"
        :disabled="!!message"
        @click.native="onSave"
      >
        Save
      </Button>
    </div>
  </div>
</template>

<script src="./index.js"></script>
<style src="./index.scss" lang="scss" module></style>
