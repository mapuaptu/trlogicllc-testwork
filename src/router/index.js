import Vue from 'vue';
import VueRouter from 'vue-router';
import Home from '../views/Home/index.vue';
import NoteView from '../views/NoteView/index.vue';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'home',
    component: Home,
  },
  {
    path: '/note/:id',
    name: 'note',
    component: NoteView,
  },
];

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});

export { routes };

export default router;
