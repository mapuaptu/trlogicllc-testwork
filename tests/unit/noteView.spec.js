import NoteView from '@/views/NoteView/index.vue';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import { routes } from '@/router';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueRouter);

const store = new Vuex.Store({
  state: {
    notes: [],
  },
});

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});

describe('Note view', () => {
  it('1. Rendered success', async () => {
    const wrapper = await mount(NoteView, { router, store, localVue });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
