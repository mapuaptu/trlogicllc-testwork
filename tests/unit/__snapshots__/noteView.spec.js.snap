// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note view 1. Rendered success 1`] = `
<div>
  <div>
    <div>

    </div>
    <div>
      <div class=""><span>
        Edit name
      </span>
        <div><svg width="24" height="24" viewbox="0 0 24 24">
            <path d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z" fill="#14b6bd"></path>
          </svg></div>
      </div>
      <div class=""><span>
        Delete note
      </span>
        <div><svg width="24" height="24" viewbox="0 0 24 24">
            <path d="M6,19A2,2 0 0,0 8,21H16A2,2 0 0,0 18,19V7H6V19M8,9H16V19H8V9M15.5,4L14.5,3H9.5L8.5,4H5V6H19V4H15.5Z" fill="#14b6bd"></path>
          </svg></div>
      </div>
    </div>
  </div>
  <div>
    <div class=""><span>
      Add todo
    </span>
      <div><svg width="24" height="24" viewbox="0 0 24 24">
          <path d="M4,9H9V4H15V9H20V15H15V20H9V15H4V9M11,13V18H13V13H18V11H13V6H11V11H6V13H11Z" fill="#fff"></path>
        </svg></div>
    </div>
    <div></div>
  </div>
  <div>
    <div class=""><span>
      Cancel
    </span>
      <!---->
    </div>
    <div class=""><span>
      Save
    </span>
      <!---->
    </div>
  </div>
</div>
`;
